devito/data/data.py

```diff
@@ -132,16 +141,82 @@ def _global(self, glb_idx, decomposition):
         ret._is_distributed = any(i is not None for i in decomposition)
         return ret
 
+    def check_slicing(func):
+        """Check if __getitem__/__setitem__ may require communication across MPI ranks."""
+        @wraps(func)
+        def wrapper(data, *args, **kwargs):
+            glb_idx = args[0]
+            if len(args) > 1 and isinstance(args[1], Data) \
+                    and args[1]._is_mpi_distributed:
+                slicing = MPI
+            elif data._is_mpi_distributed:
+                for i in as_tuple(glb_idx):
+                    if isinstance(i, slice) and i.step is not None and i.step < 0:
+                        slicing = MPI
+                        break
+                    else:
+                        slicing = STANDARD
+            else:
+                slicing = STANDARD
+            kwargs['slicing'] = slicing
+            return func(data, *args, **kwargs)
+        return wrapper
+
     @property
     def _is_mpi_distributed(self):
         return self._is_distributed and configuration['mpi']
 
     def __repr__(self):
         return super(Data, self._local).__repr__()
 
-    def __getitem__(self, glb_idx):
-        loc_idx = self._convert_index(glb_idx)
-        if loc_idx is NONLOCAL:
+    @check_slicing
+    def __getitem__(self, glb_idx, slicing):
+        loc_idx = self._convert_index_loc(glb_idx)
+        if slicing is MPI:
+            # Retrieve the pertinent local data prior to mpi send/receive operations
+            loc_data_idx = local_data_idx(loc_idx)
+            local_val = super(Data, self).__getitem__(loc_data_idx)
+
+            comm = self._distributor.comm
+            rank = self._distributor.myrank
```

Comment:
Maybe make these property if used a lot rather than access _distriubutor.somethign all the time and just do self.rank, self.comm,...

Reply:
seconded

Reply:
Sure, but since `_distributor` can be `None` these properties will need a try, except which is not normally needed since they're currently only accessed when we know `distributor` is not `None`. Can change if you'd like though?